### PR TITLE
Converter

### LINKED ISSUE #47: Add `withConverter` and pass in a generic type
# Feature request

## Is your feature request related to a problem? Please describe.

Yeah sort of, this would refer to be as "boilerplate" code. Since we typically do that manually every time we fetch data.

```dart
// where response if the returned results from Supabase client
User.fromJson(response.data)
```

## Describe the solution you'd like

Firestore has implemented this feature. So I'll just link up that from pub.dev changelog.

https://pub.dev/packages/cloud_firestore/changelog#200

## Additional context

image

### PR DESCRIPTION
**Pull request by bdlukaa**
## What kind of change does this PR introduce?

Fixes #47 

## What is the current behavior?

N/A

## What is the new behavior?

```dart
final res = await postgrest
      .from('users')
      .select()
      .withConverter<List>((data) => [data])
      .execute();
```

### CODE REVIEW

**Comment by point-source** [comment]
Thanks for working on this! I've been meaning to do it myself but just hadn't gotten to it. Definitely a useful feature. A couple of thoughts on this current proposal:

- The PostgrestResponse body will still show as dynamic even though it's been converted, right? So the user would have to cast it back into whatever type they passed to withConverter<T>
- This only handles deserialization and not serialization (since that would require making all the builders generic) so this doesn't do the same thing as firebase's withConverter method
- Should the test check to ensure that the data was converted? `expect(res.data.first, isA<List>());`

I think this is a useful contribution regardless, just trying to clarify the intended operation. Thanks

**Review by dshukertjr — COMMENTED** [review]
Thanks for working on this PR. This is amazing!

One thought I had was what if we added a generic type parameter `<T>` to `PostgrestBuilder` and any subclass of it. Could we then make it so that the return type of `execute()` will be that type? I would love to hear any thoughts on this.

**Comment by bdlukaa** [comment]
@dshukertjr Added type annotations, but I'm not sure the way it's done is correct. I think it may throw cast errors

**Comment by dshukertjr** [comment]
@bdlukaa At a quick glance, it looked amazing! Will do a through review later!

File: lib/src/postgrest_builder.dart [hunk 1]
```diff
@@ -18,12 +20,30 @@ abstract class PostgrestBuilder {
   });
 
   dynamic body;
-  final List query = [];
   final Map<String, String> headers;
   bool maybeEmpty = false;
   String? method;
   final String? schema;
   Uri url;
+  PostgrestConverter? converter;
```
Comment (dshukertjr) [inline lib/src/postgrest_builder.dart]: I think this can be private. 

```suggestion
  PostgrestConverter? _converter;
```

**Review by dshukertjr — COMMENTED** [review]
Amazing work! Very beautiful. I noticed one little thing, but other than that, it looks ready!

**Review by dshukertjr — APPROVED** [review]
LGTM! Amazing work!

**Comment by k0shk0sh** [comment]
I still don't get how this converter works? It doesn't do any kind of parsing but rather casting to list which is in most of the cases is what the response return..


Will it work if I do `List<MyAmazingClass>` because I don't think so and it can be rather misleading and will cause alot of people to figure out that there is something wrong with the library rather than their code.

Could you please answer @point-source points?

**Pull request merged**